pkg/kubeseal/kubeseal.go

```diff
@@ -0,0 +1,743 @@
+}
+
+// parseFromFile parses a value of the kubectl --from-file flag, which can optionally include an item name
+// preceding the first equals sign.
+func parseFromFile(s string) (string, string) {
+	c := strings.SplitN(s, "=", 2)
+	if len(c) == 1 {
+		return "", c[0]
+	}
+	return c[0], c[1]
+}
+
+func readPrivKeysFromFile(filename string) ([]*rsa.PrivateKey, error) {
+	// #nosec G304 -- should open user provided file
+	b, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := parsePrivKey(b)
+	if err == nil {
+		return []*rsa.PrivateKey{res}, nil
+	}
+
+	var secrets []*v1.Secret
+
+	// try to parse it as json/yaml encoded v1.List of secrets
+	var lst v1.List
+	if err = runtime.DecodeInto(scheme.Codecs.UniversalDecoder(), b, &lst); err == nil {
+		for _, r := range lst.Items {
+			s, err := readSecret(scheme.Codecs.UniversalDecoder(), bytes.NewBuffer(r.Raw))
+			if err != nil {
+				return nil, err
+			}
+			secrets = append(secrets, s)
+		}
+	} else {
+		// try to parse it as json/yaml encoded secret
+		s, err := readSecret(scheme.Codecs.UniversalDecoder(), bytes.NewBuffer(b))
+		if err != nil {
+			return nil, err
+		}
+		secrets = append(secrets, s)
+	}
+
+	var keys []*rsa.PrivateKey
+	for _, s := range secrets {
+		tlsKey, ok := s.Data["tls.key"]
+		if !ok {
+			return nil, fmt.Errorf("secret must contain a 'tls.data' key")
+		}
+		pk, err := parsePrivKey(tlsKey)
+		if err != nil {
+			return nil, err
+		}
+		keys = append(keys, pk)
+	}
+
+	return keys, nil
+}
+
+func readPrivKey(filename string) (*rsa.PrivateKey, error) {
+	pks, err := readPrivKeysFromFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	return pks[0], nil
+}
+
+func parsePrivKey(b []byte) (*rsa.PrivateKey, error) {
+	key, err := keyutil.ParsePrivateKeyPEM(b)
+	if err != nil {
+		return nil, err
+	}
+	switch rsaKey := key.(type) {
+	case *rsa.PrivateKey:
+		return rsaKey, nil
+	default:
+		return nil, fmt.Errorf("unexpected private key type %T", key)
+	}
+}
+
+func readPrivKeys(filenames []string) (map[string]*rsa.PrivateKey, error) {
+	res := map[string]*rsa.PrivateKey{}
+	for _, filename := range filenames {
+		pks, err := readPrivKeysFromFile(filename)
+		if err != nil {
+			return nil, err
+		}
+		for _, pk := range pks {
+			fingerprint, err := crypto.PublicKeyFingerprint(&pk.PublicKey)
+			if err != nil {
+				return nil, err
+			}
+
+			res[fingerprint] = pk
+		}
+	}
+	return res, nil
+}
+
+func unsealSealedSecret(flags *Flags, w io.Writer, in io.Reader, codecs runtimeserializer.CodecFactory) error {
+	privKeys, err := readPrivKeys(flags.PrivKeys)
+	if err != nil {
+		return err
+	}
+
+	b, err := io.ReadAll(in)
+	if err != nil {
+		return err
+	}
+
+	ss, err := decodeSealedSecret(codecs, b)
+	if err != nil {
+		return err
+	}
+	sec, err := ss.Unseal(codecs, privKeys)
+	if err != nil {
+		return err
+	}
+
+	return resourceOutput(w, flags, codecs, v1.SchemeGroupVersion, sec)
+}
+
+func Run(w io.Writer, cfg *Config) (err error) {
```

Comment:
I have my doubts about this function being part of the `kubeseal` pkg. IMO, the `kubeseal` pkg should provide an API that can later be used, for example, to create a CLI or a HTTP API for `kubeseal`. This function is basically running the CLI, but the `kubeseal` pkg shouldn't know anything about a CLI. It should be independent, agnostic.

Reply:
You are right, this `Run` should not be here, in fact it should not exist as is. But this PR is only a first step to move most of the `kubeseal` specific code away from the `main`, so it can be easily reused as needed.

In this shape right now you could do all the CLI does from as a library. It is not a proper library API, true, but it can be used. Whereas in the `main` it could not be reused at all from other code, only as a binary.

Note that what remains in `main` now is trully CLI dependent, whereas in `kubeseal` we have a `Flags` struct that might invoke a CLI meaning but other than that is basically not CLI aware, it does not parse flags, for instance.

Going forward, the idea is to keep refactoring and dismembering this Run into pieces. Some code will go back to `main` and probably we will get rid of the `Flags` type. Ideally, each of the behaviors of the CLI will map to functions in `kubeseal` with specific parameters, instead of a flags bag.

The plan is for smaller incremental changes, to avoid big changes that are difficult to review and may insert bugs without notice. In fact, I expect that dismembering `Run` will take more than a single PR.